docs/prompting/thought_generation/chain_of_thought_few_shot/contrastive.md

```diff
@@ -1,7 +1,134 @@
 ---
-title: ""
-description: ""
-keywords: ""
+description: "We can improve model performance by deliberating including incorrect examples of reasoning for our model to see"
 ---
 
-[wip]
+We can get better performance from our model when using chain-of-thought by including examples of incorrect reasoning. This helps our language model to learn what mistakes to avoid when generating a response. This is known as Contrastive Chain Of Thought<sup><a href="https://arxiv.org/pdf/2311.09277">1</a></sup> and can be done using the following template.
+
+!!! example "Contrastive Chain Of Thought template"
+
+    Question: **sample question**
+
+    Explanation: **correct reasoning**
+
+    Wrong Explanation: **incorrect reasoning example**
+
+    Question: **Context + Prompt**
+
+We can implement Contrastive Chain Of Thought using `instructor` as seen below.
+
+```python hl_lines="35-40"
+import instructor
+from openai import OpenAI
+from pydantic import BaseModel, Field
+from textwrap import dedent
+
+client = instructor.from_openai(OpenAI())
+
+
+class ChainOfThought(BaseModel):
+    chain_of_thought: str = Field(description="Incorrect reasoning for the answer")
+    correct_answer: str
+
+
+def contrastive_chain_of_thought(
+    query: str,
+    context: str,
+    example_prompt: str,
+    correct_examples: list[str],
+    incorrect_examples: list[str],
+):
+    correct_example_prompt = "\n".join(
+        [f"Explanation:\n{example}" for example in correct_examples]
+    )
+    incorrect_example_prompt = "\n".join(
+        [f"Wrong Explanation:\n{example}" for example in incorrect_examples]
+    )
+    ""
+    return client.chat.completions.create(
+        model="gpt-4o",
+        response_model=ChainOfThought,
+        messages=[
+            {
+                "role": "system",
+                "content": dedent(f"""
+            You are an expert question answering AI System.
+
+            You are about to be given some examples of incorrect
+            and correct reasoning for a question. You will then
+            be asked to correctly reason through another question
+            to generate a valid response.
+
+            Question: {example_prompt}
+
+            {correct_example_prompt}
```

Comment:
we should just start using xml when possible rather than plan string formatting.